src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -698,4 +570,260 @@ public SNIError GetLastError()
+        internal int Port { get; private set; } = -1;
+
+        public string InstanceName { get; internal set; }
+
+        public string PipeName { get; internal set; }
+
+
+        private string _workingDataSource;
+        private string _dataSourceAfterTrimmingProtocol;
+        internal bool IsBadDataSource { get; private set; } = false;
+
+        private DataSource(string dataSource)
+        {
+            // Remove all whitespaces from the datasource and all operations will happen on lower case.
+            _workingDataSource = dataSource.Trim().ToLower();
+
+            int firstIndexOfColon = _workingDataSource.IndexOf(':');
+
+            PopulateProtocol();
+
+            _dataSourceAfterTrimmingProtocol = (firstIndexOfColon > -1) && ConnectionProtocol != DataSource.Protocol.None
+                ? _workingDataSource.Substring(firstIndexOfColon + 1).Trim() : _workingDataSource;
+
+            if (_dataSourceAfterTrimmingProtocol.Contains("/")) // Pipe paths only allow back slashes
+            {
+                if (ConnectionProtocol == DataSource.Protocol.None)
+                    ReportSNIError(SNIProviders.INVALID_PROV);
+                else if (ConnectionProtocol == DataSource.Protocol.NP)
+                    ReportSNIError(SNIProviders.NP_PROV);
+                else if (ConnectionProtocol == DataSource.Protocol.TCP)
+                    ReportSNIError(SNIProviders.TCP_PROV);
+            }
+        }
+
+        private void PopulateProtocol()
+        {
+            string[] splitByColon = _workingDataSource.Split(':');
+
+            if (splitByColon.Length <= 1)
+            {
+                ConnectionProtocol = DataSource.Protocol.None;
+            }
+            else
+            {
+                // We trim before switching because " tcp : server , 1433 " is a valid data source
+                switch (splitByColon[0].Trim())
+                {
+                    case TdsEnums.TCP:
+                        ConnectionProtocol = DataSource.Protocol.TCP;
+                        break;
+                    case TdsEnums.NP:
+                        ConnectionProtocol = DataSource.Protocol.NP;
+                        break;
+                    case TdsEnums.ADMIN:
+                        ConnectionProtocol = DataSource.Protocol.Admin;
+                        break;
+                    default:
+                        // None of the supported protocols were found. This may be a IPv6 address
+                        ConnectionProtocol = DataSource.Protocol.None;
+                        break;
+                }
+            }
+        }
+
+        public static DataSource ParseServerName(string dataSource)
+        {
+            DataSource details = new DataSource(dataSource);
+
+            if (details.IsBadDataSource)
+            {
+                return null;
+            }
+
+            if (details.InferNamedPipesInformation() && !details.IsBadDataSource)
```

Comment:
Since IsBadDataSource is only set on failure, do we need extra checks for it in these if cases?